From 11189ea7e7b3a3ebe67e3c22ac03799e960b27a9 Mon Sep 17 00:00:00 2001
From: Andrei Bacescu <9010245+bacescuandrei@users.noreply.github.com>
Date: Tue, 27 Oct 2026 23:08:54 +0200
Subject: [PATCH] 10043: RISC-V: add p-code for the Zba instructions

sh1add, sh2add, sh3add, add.uw, sh1add.uw, sh2add.uw, sh3add.uw and
slli.uw (addu.w, shNaddu.w and slliu.w in the spec) were decoded but
marked unimpl, so the decompiler could not follow address computations
that compilers emit with Zba.
---
 .../RISCV/data/languages/riscv.rv32b.sinc     | 19 ++++++++----
 .../RISCV/data/languages/riscv.rv64b.sinc     | 30 +++++++++++++++----
 2 files changed, 39 insertions(+), 10 deletions(-)

diff --git a/Ghidra/Processors/RISCV/data/languages/riscv.rv32b.sinc b/Ghidra/Processors/RISCV/data/languages/riscv.rv32b.sinc
index 46852a2aead..737e47a31e4 100644
--- a/Ghidra/Processors/RISCV/data/languages/riscv.rv32b.sinc
+++ b/Ghidra/Processors/RISCV/data/languages/riscv.rv32b.sinc
@@ -101,11 +101,20 @@
 
 :sext.h rd, rs1 is op0006=0x13 & op1214=0x1 & op2024=0x5 & op2531=0x30 & rd & rs1 unimpl
 
-:sh1add rd, rs1, rs2 is op0006=0x33 & op1214=0x2 & op2531=0x10 & rd & rs1 & rs2 unimpl
-
-:sh2add rd, rs1, rs2 is op0006=0x33 & op1214=0x4 & op2531=0x10 & rd & rs1 & rs2 unimpl
-
-:sh3add rd, rs1, rs2 is op0006=0x33 & op1214=0x6 & op2531=0x10 & rd & rs1 & rs2 unimpl
+:sh1add rd, rs1, rs2 is op0006=0x33 & op1214=0x2 & op2531=0x10 & rd & rs1 & rs2
+{
+	rd = (rs1 << 1) + rs2;
+}
+
+:sh2add rd, rs1, rs2 is op0006=0x33 & op1214=0x4 & op2531=0x10 & rd & rs1 & rs2
+{
+	rd = (rs1 << 2) + rs2;
+}
+
+:sh3add rd, rs1, rs2 is op0006=0x33 & op1214=0x6 & op2531=0x10 & rd & rs1 & rs2
+{
+	rd = (rs1 << 3) + rs2;
+}
 
 :shfl    rd, rs1, rs2 is op0006=0x33 & op1214=0x1 & op2531=0x4 & rd & rs1 & rs2 unimpl
 
diff --git a/Ghidra/Processors/RISCV/data/languages/riscv.rv64b.sinc b/Ghidra/Processors/RISCV/data/languages/riscv.rv64b.sinc
index e836096fcd5..44d100e65f7 100644
--- a/Ghidra/Processors/RISCV/data/languages/riscv.rv64b.sinc
+++ b/Ghidra/Processors/RISCV/data/languages/riscv.rv64b.sinc
@@ -3,7 +3,11 @@
 #TODO  fix op2031
 :addiwu rd, rs1, op2031 is op0006=0x1b & op1214=0x4 & rd & op2031 & rs1 unimpl
 
-:addu.w rd, rs1, rs2 is op0006=0x3b & op1214=0x0 & op2531=0x4 & rd & rs1 & rs2 unimpl
+:addu.w rd, rs1, rs2 is op0006=0x3b & op1214=0x0 & op2531=0x4 & rd & rs1 & rs2
+{
+	local tmp:$(XLEN) = zext(rs1:4);
+	rd = tmp + rs2;
+}
 
 :addwu rd, rs1, rs2 is op0006=0x3b & op1214=0x0 & op2531=0x5 & rd & rs1 & rs2 unimpl
 
@@ -80,16 +84,32 @@
 
 :sbsetw  rd, rs1, rs2 is op0006=0x3b & op1214=0x1 & op2531=0x14 & rd & rs1 & rs2 unimpl
 
-:sh1addu.w rd, rs1, rs2 is op0006=0x3b & op1214=0x2 & op2531=0x10 & rd & rs1 & rs2 unimpl
+:sh1addu.w rd, rs1, rs2 is op0006=0x3b & op1214=0x2 & op2531=0x10 & rd & rs1 & rs2
+{
+	local tmp:$(XLEN) = zext(rs1:4);
+	rd = (tmp << 1) + rs2;
+}
 
-:sh2addu.w rd, rs1, rs2 is op0006=0x3b & op1214=0x4 & op2531=0x10 & rd & rs1 & rs2 unimpl
+:sh2addu.w rd, rs1, rs2 is op0006=0x3b & op1214=0x4 & op2531=0x10 & rd & rs1 & rs2
+{
+	local tmp:$(XLEN) = zext(rs1:4);
+	rd = (tmp << 2) + rs2;
+}
 
-:sh3addu.w rd, rs1, rs2 is op0006=0x3b & op1214=0x6 & op2531=0x10 & rd & rs1 & rs2 unimpl
+:sh3addu.w rd, rs1, rs2 is op0006=0x3b & op1214=0x6 & op2531=0x10 & rd & rs1 & rs2
+{
+	local tmp:$(XLEN) = zext(rs1:4);
+	rd = (tmp << 3) + rs2;
+}
 
 :shflw    rd, rs1, rs2 is op0006=0x3b & op1214=0x1 & op2531=0x4 & rd & rs1 & rs2 unimpl
 
 #TODO  fix op2026
-:slliu.w rd, rs1, op2026 is op0006=0x1b & op1214=0x1 & op2731=0x1 & op2026 & rd & rs1 unimpl
+:slliu.w rd, rs1, op2026 is op0006=0x1b & op1214=0x1 & op2731=0x1 & op2026 & rd & rs1
+{
+	local tmp:$(XLEN) = zext(rs1:4);
+	rd = tmp << op2026;
+}
 
 #TODO  fix op2024
 :sloiw rd, rs1, op2024 is op0006=0x1b & op1214=0x1 & op2531=0x10 & op2024 & rd & rs1 unimpl
